File: construction_management/construction_management/page/project_process_home/project_process_home.js
// Copyright (c) 2026, Construction Management
// License: MIT

frappe.pages['project-process-home'].on_page_load = function (wrapper) {
	frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Member Home'),
		single_column: true,
	});

	new ProjectProcessHome(wrapper);
};

class ProjectProcessHome {
	project_id(value) {
		return String(value == null ? '' : value);
	}

	constructor(wrapper) {
		this.wrapper = $(wrapper);
		this.page_body = this.wrapper.find('.layout-main-section');
		this.status_filter = 'ongoing';
		this.search = '';
		this.company = frappe.defaults.get_user_default('Company') || '';
		this.start = 0;
		this.page_length = 25;
		this.total_count = 0;
		this.expanded_projects = new Set();
		this.process_cache = {};
		this.bills_cache = {};
		this.boq_source_cache = {};
		this.loading_projects = new Set();
		this.projects = [];
		this.companies = [];

		this.render_shell();
		this.bind_events();
		this.load_companies();
		this.load_projects();
	}

	render_shell() {
		this.page_body.html(`
			<div class="project-process-home-layout">
				<div class="pph-toolbar">
					<div class="pph-status-filters">
						<label><input type="radio" name="pph_status" value="ongoing" checked> ${__('On Going')}</label>
						<label><input type="radio" name="pph_status" value="completed"> ${__('Completed')}</label>
						<label><input type="radio" name="pph_status" value="all"> ${__('All')}</label>
					</div>
					<div class="pph-company-filter">
						<label>${__('Company')}</label>
						<select class="form-control input-sm" data-field="company" style="width: 220px; display: inline-block;">
							<option value="">${__('All Companies')}</option>
						</select>
					</div>
					<div class="pph-search-filter">
						<label>${__('Search')}</label>
						<div class="pph-search-input-wrap">
							<input
								type="text"
								class="form-control input-sm"
								data-field="search"
								placeholder="${__('Project no, name, engineer, location...')}"
								value="${frappe.utils.escape_html(this.search || '')}"
							>
							<button type="button" class="btn btn-default btn-xs pph-search-clear" data-action="clear-search" title="${__('Clear')}">&times;</button>
						</div>
					</div>
					<div class="pph-toolbar-actions">
						<button class="btn btn-default btn-sm" data-action="collapse-all">${__('Collapse All')}</button>
						<button class="btn btn-default btn-sm" data-action="expand-all">${__('Expand All')}</button>
					</div>
				</div>
				<div class="pph-table-controls">
					<div class="entries-control">
						<span>${__('Show')}</span>
						<select class="form-control input-sm" data-field="page-length" style="width: 70px; display: inline-block;">
							<option value="25">25</option>
							<option value="50">50</option>
							<option value="100">100</option>
						</select>
						<span>${__('entries')}</span>
					</div>
				</div>
				<div class="pph-table-wrapper">
					<table class="pph-table">
						<thead>
							<tr class="pph-parent-header">
								<th>${__('Project_No.')}</th>
								<th>${__('Sales Person')}</th>
								<th>${__('Assign_To')}</th>
								<th>${__('Contractor')}</th>
								<th>${__('Proname')}</th>
								<th>${__('Location')}</th>
								<th>${__('emirates')}</th>
							</tr>
						</thead>
						<tbody id="pph-project-tbody"></tbody>
					</table>
				</div>
				<div class="pph-pagination">
					<div class="pph-pagination-info"></div>
					<div class="btn-group">
						<button class="btn btn-default btn-sm" data-action="prev">${__('Previous')}</button>
						<button class="btn btn-default btn-sm" data-action="next">${__('Next')}</button>
					</div>
				</div>
			</div>
		`);
	}

	bind_events() {
		const me = this;
		this.page_body.on('change', 'input[name="pph_status"]', function () {
			me.status_filter = $(this).val();
			me.start = 0;
			me.expanded_projects.clear();
			me.load_projects();
		});

		this.page_body.on('change', '[data-field="page-length"]', function () {
			me.page_length = parseInt($(this).val(), 10) || 25;
			me.start = 0;
			me.load_projects();
		});

		this.page_body.on('change', '[data-field="company"]', function () {
			me.company = $(this).val();
			me.start = 0;
			me.expanded_projects.clear();
			me.process_cache = {};
			me.bills_cache = {};
			me.boq_source_cache = {};
			me.load_projects();
		});

		let search_timer = null;
		this.page_body.on('input', '[data-field="search"]', function () {
			clearTimeout(search_timer);
			search_timer = setTimeout(() => {
				me.search = $(this).val();
				me.start = 0;
				me.expanded_projects.clear();
				me.load_projects();
			}, 300);
		});

		this.page_body.on('keydown', '[data-field="search"]', function (e) {
			if (e.key === 'Enter') {
				e.preventDefault();
				clearTimeout(search_timer);
				me.search = $(this).val();
				me.start = 0;
				me.expanded_projects.clear();
				me.load_projects();
			}
		});

		this.page_body.on('click', '[data-action="clear-search"]', () => {
			me.search = '';
			me.start = 0;
			me.expanded_projects.clear();
			me.page_body.find('[data-field="search"]').val('');
			me.load_projects();
		});

		this.page_body.on('click', '[data-action="prev"]', () => {
			if (me.start > 0) {
				me.start = Math.max(0, me.start - me.page_length);
				me.load_projects();
			}
		});

		this.page_body.on('click', '[data-action="next"]', () => {
			if (me.start + me.page_length < me.total_count) {
				me.start += me.page_length;
				me.load_projects();
			}
		});

		this.page_body.on('click', '[data-action="collapse-all"]', () => {
			me.expanded_projects.clear();
			me.render_project_rows();
		});

		this.page_body.on('click', '[data-action="expand-all"]', () => {
			me.projects.forEach((p) => me.expanded_projects.add(me.project_id(p.name)));
			me.render_project_rows();
			me.projects.forEach((p) => {
				const project_name = me.project_id(p.name);
				if (me.process_cache[project_name] === undefined && !me.loading_projects.has(project_name)) {
					me.load_process_rows(project_name);
				}
			});
		});

		this.page_body.on('click', '.pph-expand-btn', function (e) {
			e.stopPropagation();
			const project = me.project_id($(this).attr('data-project'));
			me.toggle_expand(project);
		});

		this.page_body.on('click', '.pph-project-no', function (e) {
			e.stopPropagation();
			const project = me.project_id($(this).attr('data-project'));
			frappe.set_route('Form', 'Project', project);
		});

		this.page_body.on('click', '.pph-edit-process', function (e) {
			e.stopPropagation();
			const doctype = $(this).data('doctype');
			const docname = $(this).data('docname');
			if (doctype && docname) {
				frappe.set_route('Form', doctype, docname);
			}
		});

		this.page_body.on('click', '.pph-add-process', function (e) {
			e.stopPropagation();
			const project = me.project_id($(this).attr('data-project'));
			me.show_add_process_dialog(project);
		});
		this.page_body.on('click', '.pph-open-project', function (e) {
			e.stopPropagation();
			frappe.set_route('Form', 'Project', me.project_id($(this).attr('data-project')));
		});

		this.page_body.on('click', '.pph-open-boq', function (e) {
			e.stopPropagation();
			const project = me.project_id($(this).attr('data-project'));
			frappe.set_route('Form', 'Project', project, { tab: 'construction_tab' });
		});
	}

	load_companies() {
		const me = this;
		frappe.call({
			method: 'construction_management.construction_management.page.project_process_home.project_process_home.get_member_home_companies',
			callback(r) {
				if (!r.message) {
					return;
				}
				me.companies = r.message;
				const $select = me.page_body.find('[data-field="company"]');
				let options = `<option value="">${__('All Companies')}</option>`;
				me.companies.forEach((company) => {
					const selected = company.name === me.company ? 'selected' : '';
					options += `<option value="${frappe.utils.escape_html(company.name)}" ${selected}>${frappe.utils.escape_html(company.company_name || company.name)}</option>`;
				});
				$select.html(options);
			},
		});
	}

	load_projects() {
		const me = this;
		frappe.call({
			method: 'construction_management.construction_management.page.project_process_home.project_process_home.get_project_process_home_data',
			args: {
				status_filter: this.status_filter,
				search: this.search,
				company: this.company,
				start: this.start,
				page_length: this.page_length,
			},
			callback(r) {
				if (r.message) {
					me.projects = r.message.projects || [];
					me.total_count = r.message.total_count || 0;
					me.render_project_rows();
					me.render_pagination();
				}
			},
		});
	}

	render_pagination() {
		const from = this.total_count ? this.start + 1 : 0;
		const to = Math.min(this.start + this.page_length, this.total_count);
		this.page_body.find('.pph-pagination-info').text(
			__('Showing {0} to {1} of {2} entries', [from, to, this.total_count])
		);
		this.page_body.find('[data-action="prev"]').prop('disabled', this.start <= 0);
		this.page_body.find('[data-action="next"]').prop(
			'disabled',
			this.start + this.page_length >= this.total_count
		);
	}

	render_project_rows() {
		const tbody = this.page_body.find('#pph-project-tbody');
		if (!this.projects.length) {
			tbody.html(`
				<tr class="pph-empty-row">
					<td colspan="7">${__('No projects found')}</td>
				</tr>
			`);
			return;
		}

		let html = '';
		this.projects.forEach((project) => {
			const project_name = this.project_id(project.name);
			const expanded = this.expanded_projects.has(project_name);
			const assign_class = project.assign_to === __('!Not Assign') ? 'pph-not-assigned' : '';
			html += `
				<tr class="pph-project-row" data-project="${frappe.utils.escape_html(project_name)}">
					<td>
						<button type="button" class="pph-expand-btn ${expanded ? 'expanded' : ''}" data-project="${frappe.utils.escape_html(project_name)}">
							${expanded ? '−' : '+'}
						</button>
						<span class="pph-project-no" data-project="${frappe.utils.escape_html(project_name)}">${frappe.utils.escape_html(project.project_no)}</span>
					</td>
					<td>${frappe.utils.escape_html(project.sales_person || '')}</td>
					<td class="${assign_class}">${frappe.utils.escape_html(project.assign_to || '')}</td>
					<td>${frappe.utils.escape_html(project.contractor || '')}</td>
					<td>${frappe.utils.escape_html(project.project_name || '')}</td>
					<td>${frappe.utils.escape_html(project.location || '')}</td>
					<td>${frappe.utils.escape_html(project.emirates || '')}</td>
				</tr>
			`;
			if (expanded) {
				html += this.render_process_section(project_name);
			}
		});
		tbody.html(html);
	}

	render_process_section(project) {
		const cached = this.process_cache[project];
		const bills = this.bills_cache[project] || [];
		const boq_source_project = this.boq_source_cache[project];
		const project_row = this.projects.find((p) => p.name === project) || {};
		const enable_progressive_boq = project_row.enable_progressive_boq ? 1 : 0;

		if (cached === undefined) {
			return `
				<tr class="pph-process-section" data-project="${project}">
					<td colspan="7" class="pph-loading-row">${__('Loading processes...')}</td>
				</tr>
			`;
		}

		let note_html = '';
		if (boq_source_project && boq_source_project !== project) {
			note_html = `
				<tr class="pph-boq-source-note">
					<td colspan="7" class="text-muted">
						${__('Showing BOQ from {0}', [boq_source_project])}
					</td>
				</tr>
			`;
		}

		let rows_html = '';
		if (cached.length) {
			cached.forEach((row) => {
				const area_disp = frappe.format(row.area, { fieldtype: 'Float', precision: 2 });
				const skirting_disp = `${frappe.format(row.skirting || 0, { fieldtype: 'Float', precision: 2 })} ${row.unit || 'm²'}`;
				const start_disp = row.start_date ? frappe.datetime.str_to_user(row.start_date) : '';
				const end_disp = row.end_date ? frappe.datetime.str_to_user(row.end_date) : '';

				rows_html += `
					<tr class="pph-process-row">
						<td>${frappe.utils.escape_html(row.process_name || '')}</td>
						<td>${area_disp}</td>
						<td>${skirting_disp}</td>
						<td>${start_disp}</td>
						<td>${end_disp}</td>
						<td><span class="pph-action-link pph-edit-process" data-doctype="BOQ Item" data-docname="${row.boq_item}">${__('Edit')}</span></td>
						<td></td>
					</tr>
				`;
			});
		} else {
			const boq_btn = enable_progressive_boq
				? `<button class="btn btn-default btn-xs pph-open-boq" data-project="${project}">${__('Open Construction BOQ')}</button>`
				: '';
			rows_html = `
				<tr>
					<td colspan="7" class="pph-empty-boq text-center text-muted">
						<p>${__('No BOQ items configured for this project')}</p>
						<div class="pph-empty-actions">
							<button class="btn btn-default btn-xs pph-open-project" data-project="${project}">${__('Open Project')}</button>
							${boq_btn}
						</div>
					</td>
				</tr>
			`;
		}

		if (bills.length) {
			rows_html += `
				<tr class="pph-process-row">
					<td colspan="6"></td>
					<td>
						<span class="pph-action-link pph-add-process" data-project="${project}">
							${__('Add New')}
						</span>
					</td>
				</tr>
			`;
		}

		return `
			<tr class="pph-process-section" data-project="${project}">
				<td colspan="7" style="padding: 0;">
					<table class="pph-table" style="margin: 0;">
						<thead>
							<tr class="pph-child-header">
								<th>${__('Process Name')}</th>
								<th>${__('Area')}</th>
								<th>${__('Skirting')}</th>
								<th>${__('Start Date')}</th>
								<th>${__('End Date')}</th>
								<th>${__('Edit')}</th>
								<th>${__('Add New')}</th>
							</tr>
						</thead>
						<tbody>${note_html}${rows_html}</tbody>
					</table>
				</td>
			</tr>
		`;
	}

	toggle_expand(project) {
		project = this.project_id(project);
		if (this.expanded_projects.has(project)) {
			this.expanded_projects.delete(project);
		} else {
			this.expanded_projects.add(project);
			if (this.process_cache[project] === undefined && !this.loading_projects.has(project)) {
				this.load_process_rows(project);
			}
		}
		this.render_project_rows();
	}

	load_process_rows(project) {
		const me = this;
		project = me.project_id(project);
		me.loading_projects.add(project);
		if (me.expanded_projects.has(project)) {
			me.render_project_rows();
		}
		frappe.call({
			method: 'construction_management.construction_management.page.project_process_home.project_process_home.get_project_process_rows',
			args: { project },
			callback(r) {
				me.loading_projects.delete(project);
				if (r.message) {
					me.process_cache[project] = r.message.processes || [];
					me.bills_cache[project] = r.message.bills || [];
					me.boq_source_cache[project] = r.message.boq_source_project || project;
					if (me.expanded_projects.has(project)) {
						me.render_project_rows();
					}
				}
			},
			error() {
				me.loading_projects.delete(project);
				me.process_cache[project] = [];
				me.bills_cache[project] = [];
				if (me.expanded_projects.has(project)) {
					me.render_project_rows();
				}
			},
		});
	}

	show_add_process_dialog(project) {
		const bills = this.bills_cache[project] || [];
		if (!bills.length) {
			frappe.msgprint({
				title: __('No BOQ Bill'),
				message: __('This project has no BOQ Bill. Please create a BOQ Bill first from the Project Construction tab.'),
				indicator: 'orange',
			});
			return;
		}

		const bill_options = bills.map((bill) => {
			const label = bill.bill_no || bill.label || bill.name;
			return { label, value: bill.name };
		});

		const me = this;
		const dialog = new frappe.ui.Dialog({
			title: __('Add New BOQ Item'),
			fields: [
				{
					fieldname: 'parent_bill',
					label: __('BOQ Bill'),
					fieldtype: 'Select',
					options: bill_options,
					reqd: 1,
					default: bill_options[0]?.value,
				},
				{ fieldname: 'description', label: __('Process Name'), fieldtype: 'Data', reqd: 1 },
				{ fieldname: 'total_qty', label: __('Area'), fieldtype: 'Float' },
				{ fieldname: 'custom_skirting', label: __('Skirting'), fieldtype: 'Float' },
				{ fieldname: 'start_date', label: __('Start Date'), fieldtype: 'Date' },
				{ fieldname: 'end_date', label: __('End Date'), fieldtype: 'Date' },
				{ fieldname: 'rate', label: __('Rate'), fieldtype: 'Currency', default: 0 },
			],
			primary_action_label: __('Create'),
			primary_action(values) {
				frappe.call({
					method: 'construction_management.construction_management.page.project_process_home.project_process_home.create_project_process_boq_item',
					args: {
						project,
						parent_bill: values.parent_bill,
						description: values.description,
						total_qty: values.total_qty || 0,
						custom_skirting: values.custom_skirting || 0,
						start_date: values.start_date,
						end_date: values.end_date,
						rate: values.rate || 0,
					},
					callback(r) {
						if (r.message) {
							dialog.hide();
							delete me.process_cache[project];
							me.load_process_rows(project);
							frappe.show_alert({ message: __('BOQ Item created'), indicator: 'green' });
						}
					},
				});
			},
		});
		dialog.show();
	}
}

File: construction_management/construction_management/page/project_process_home/project_process_home.py
# Copyright (c) 2026, Construction Management
# License: MIT

import frappe
from frappe import _
from frappe.utils import cint, flt

from construction_management.api.boq_tasks import create_boq_item_with_task
from construction_management.construction_management.doctype.project_tab_access.project_tab_access import (
	get_user_project_scope,
)
from construction_management.overrides.project import get_project_contractor_name


STATUS_FILTERS = {
	"ongoing": ("Open",),
	"completed": ("Completed",),
	"all": None,
}


def _active_project_filters() -> dict:
	"""ERPNext Project.is_active is Yes/No, not 0/1."""
	return {"is_active": "Yes"}


def _apply_member_project_scope(filters: dict) -> list[str] | None:
	"""Apply Project Access selections to Member Home data queries."""
	scope = get_user_project_scope(frappe.session.user)
	if scope is not None:
		filters["name"] = ["in", scope]
	return scope


def _boq_item_optional_fields() -> list[str]:
	fields = []
	for fieldname in ("start_date", "end_date", "custom_skirting"):
		if frappe.db.exists("Custom Field", {"dt": "BOQ Item", "fieldname": fieldname}):
			fields.append(fieldname)
	return fields


@frappe.whitelist()
def get_project_process_home_data(
	status_filter: str = "ongoing",
	search: str = "",
	company: str = "",
	start: int = 0,
	page_length: int = 25,
	project_number_sort: str = "asc",
) -> dict:
	"""Return paginated project list for the member home page."""
	status_filter = (status_filter or "ongoing").lower()
	if status_filter not in STATUS_FILTERS:
		frappe.throw(_("Invalid status filter"))

	start = cint(start)
	page_length = max(1, min(cint(page_length) or 25, 100))
	project_number_sort = (project_number_sort or "asc").lower()
	if project_number_sort not in {"asc", "desc"}:
		frappe.throw(_("Project number sort must be ascending or descending"))

	filters = _active_project_filters()
	_apply_member_project_scope(filters)
	statuses = STATUS_FILTERS[status_filter]
	if statuses:
		filters["status"] = ["in", list(statuses)]

	if company:
		filters["company"] = company

	projects = frappe.get_all(
		"Project",
		filters=filters,
		fields=[
			"name",
			"company",
			"custom_project_no",
			"custom_sales_engineer",
			"custom_project_engineer",
			"customer",
			"contractor",
			"project_name",
			"custom_location",
			"custom_emirates",
			"status",
			"enable_progressive_boq",
		],
		order_by=f"custom_project_no {project_number_sort}, name {project_number_sort}",
	)

	employee_ids = {
		p.custom_sales_engineer for p in projects if p.custom_sales_engineer
	} | {p.custom_project_engineer for p in projects if p.custom_project_engineer}
	employee_names = _get_employee_names(list(employee_ids))
	contractor_names = {
		p.name: get_project_contractor_name(p)
		for p in projects
		if getattr(p, "customer", None) or getattr(p, "contractor", None)
	}

	if search:
		needle = search.strip().lower()
		projects = [
			p
			for p in projects
			if _project_matches_search(p, needle, employee_names, contractor_names)
		]

	total_count = len(projects)
	page = projects[start : start + page_length]

	rows = []
	for project in page:
		rows.append(
			{
				"name": project.name,
				"company": project.company or "",
				"project_no": project.custom_project_no or project.name,
				"sales_person": employee_names.get(project.custom_sales_engineer, ""),
				"assign_to": employee_names.get(project.custom_project_engineer, _("!Not Assign")),
				"contractor": contractor_names.get(project.name, ""),
				"project_name": project.project_name,
				"location": project.custom_location or "",
				"emirates": project.custom_emirates or "",
				"status": project.status,
				"enable_progressive_boq": cint(project.enable_progressive_boq),
			}
		)

	return {
		"projects": rows,
		"total_count": total_count,
		"start": start,
		"page_length": page_length,
	}


@frappe.whitelist()
def get_member_home_companies() -> list[dict]:
	"""Distinct companies from active projects for the member home filter."""
	filters = _active_project_filters()
	_apply_member_project_scope(filters)
	project_companies = frappe.get_all(
		"Project",
		filters=filters,
		fields=["company"],
		group_by="company",
		order_by="company asc",
	)
	company_names = [row.company for row in project_companies if row.company]
	if not company_names:
		return []

	companies = frappe.get_all(
		"Company",
		filters={"name": ["in", company_names]},
		fields=["name", "company_name"],
		order_by="company_name asc, name asc",
	)
	return [{"name": row.name, "company_name": row.company_name or row.name} for row in companies]


def _project_matches_search(project, needle: str, employee_names: dict, contractor_names: dict) -> bool:
	search_values = [
		project.custom_project_no,
		project.project_name,
		project.name,
		project.custom_location,
		project.custom_emirates,
		project.custom_sales_engineer,
		project.custom_project_engineer,
		project.customer,
		project.contractor,
		employee_names.get(project.custom_sales_engineer),
		employee_names.get(project.custom_project_engineer),
		contractor_names.get(project.name),
	]
	return any(needle in (value or "").lower() for value in search_values)


def _installment_exclusion_sql() -> str:
	"""Exclude payment-milestone BOQ rows (e.g. 1st Installment) from scope/process list."""
	return """
		AND LOWER(COALESCE(bi.description, '')) NOT LIKE '%%installment%%'
		AND LOWER(COALESCE(bi.item_code, '')) NOT LIKE '%%installment%%'
		AND LOWER(COALESCE(bi.label, '')) NOT LIKE '%%installment%%'
	"""


def _get_related_construction_project(project: str) -> str | None:
	"""Map billing-side projects (e.g. SKD-46 (SIDE)) to main construction project BOQ."""
	candidates = []
	if "(SIDE)" in project:
		candidates.append(project.replace("(SIDE)", "").strip())

	project_no = frappe.db.get_value("Project", project, "custom_project_no") or ""
	if project_no and "SIDE" in project_no.upper():
		base_no = project_no.split("(")[0].strip()
		match = frappe.db.get_value("Project", {"custom_project_no": base_no}, "name")
		if match:
			candidates.append(match)

	for candidate in candidates:
		if candidate and candidate != project and frappe.db.exists("Project", candidate):
			if _has_scope_boq_items(candidate):
				return candidate
	return None


def _has_scope_boq_items(project: str) -> bool:
	return bool(
		frappe.db.sql(
			f"""
			SELECT 1
			FROM `tabBOQ Item` bi
			WHERE bi.project = %s
			{_installment_exclusion_sql()}
			LIMIT 1
			""",
			project,
		)
	)


def _resolve_boq_source_project(project: str) -> str:
	if _has_scope_boq_items(project):
		return project
	related = _get_related_construction_project(project)
	return related or project


def _fetch_scope_boq_items(project: str) -> list[dict]:
	optional_fields = _boq_item_optional_fields()
	extra_select = "".join(f", bi.`{field}`" for field in optional_fields)

	return frappe.db.sql(
		f"""
		SELECT
			bi.name,
			bi.description,
			bi.label,
			bi.item_code,
			bi.total_qty,
			bi.unit,
			bi.parent_bill,
			bb.bill_no,
			bb.label AS bill_label,
			bb.sequence
			{extra_select}
		FROM `tabBOQ Item` bi
		INNER JOIN `tabBOQ Bill` bb ON bb.name = bi.parent_bill
		WHERE bi.project = %s
		{_installment_exclusion_sql()}
		ORDER BY bb.sequence ASC, bi.idx ASC, bi.creation ASC
		""",
		project,
		as_dict=True,
	)


@frappe.whitelist()
def get_project_process_rows(project: str) -> dict:
	"""Return construction BOQ Bill + BOQ Item rows (excludes payment installments)."""
	scope = get_user_project_scope(frappe.session.user)
	if scope is not None and project not in scope:
		frappe.throw(_("You do not have access to this project."), frappe.PermissionError)
	if not project:
		frappe.throw(_("Project is required"))

	if not frappe.db.exists("Project", project):
		frappe.throw(_("Project {0} does not exist").format(project))

	source_project = _resolve_boq_source_project(project)
	rows = _fetch_scope_boq_items(source_project)
	process_rows = [_boq_item_as_process_row(row) for row in rows]
	bills = _get_scope_boq_bills(source_project)

	return {
		"project": project,
		"boq_source_project": source_project,
		"processes": process_rows,
		"bills": bills,
	}


@frappe.whitelist()
def create_project_process_boq_item(
	project: str,
	parent_bill: str,
	description: str,
	total_qty: float = 0,
	unit: str = "m²",
	rate: float = 0,
	start_date: str = None,
	end_date: str = None,
	custom_skirting: float = 0,
) -> dict:
	"""Create a BOQ Item (process row) under a BOQ Bill — no ERPNext Task."""
	if not project or not parent_bill or not description:
		frappe.throw(_("Project, BOQ Bill, and Process Name are required"))

	bill_project = frappe.db.get_value("BOQ Bill", parent_bill, "project")
	source_project = _resolve_boq_source_project(project)
	if bill_project != source_project:
		frappe.throw(_("Selected BOQ Bill does not belong to this project's BOQ"))

	result = create_boq_item_with_task(
		parent_bill=parent_bill,
		description=description,
		unit=unit or "m²",
		total_qty=flt(total_qty),
		rate=flt(rate),
		is_task=0,
		start_date=start_date,
		end_date=end_date,
	)

	boq_item_name = result.get("boq_item")
	if boq_item_name and frappe.db.exists("Custom Field", {"dt": "BOQ Item", "fieldname": "custom_skirting"}):
		frappe.db.set_value("BOQ Item", boq_item_name, "custom_skirting", flt(custom_skirting))

	return result


def _get_scope_boq_bills(project: str) -> list[dict]:
	"""Return BOQ Bills that contain at least one non-installment scope item."""
	return frappe.db.sql(
		f"""
		SELECT DISTINCT
			bb.name,
			bb.bill_no,
			bb.label,
			bb.sequence
		FROM `tabBOQ Bill` bb
		INNER JOIN `tabBOQ Item` bi ON bi.parent_bill = bb.name
		WHERE bb.project = %s
		{_installment_exclusion_sql()}
		ORDER BY bb.sequence ASC, bb.creation ASC
		""",
		project,
		as_dict=True,
	)


def _get_project_boq_bills(project: str) -> list[dict]:
	return _get_scope_boq_bills(project)


def _boq_item_as_process_row(boq_item: dict) -> dict:
	start_date = boq_item.get("start_date")
	end_date = boq_item.get("end_date")

	return {
		"process_name": boq_item.get("description") or boq_item.get("label") or boq_item.get("name"),
		"area": flt(boq_item.get("total_qty")),
		"skirting": flt(boq_item.get("custom_skirting")),
		"unit": boq_item.get("unit") or "m²",
		"start_date": str(start_date) if start_date else None,
		"end_date": str(end_date) if end_date else None,
		"boq_item": boq_item.get("name"),
		"boq_bill": boq_item.get("parent_bill"),
		"bill_no": boq_item.get("bill_no") or "",
		"bill_label": boq_item.get("bill_label") or "",
		"doctype": "BOQ Item",
	}


def _get_employee_names(employee_ids: list) -> dict:
	employee_ids = list({e for e in employee_ids if e})
	if not employee_ids:
		return {}
	rows = frappe.get_all(
		"Employee",
		filters={"name": ["in", employee_ids]},
		fields=["name", "employee_name"],
	)
	return {row.name: row.employee_name for row in rows}


def _get_supplier_names(supplier_ids: list) -> dict:
	supplier_ids = list({s for s in supplier_ids if s})
	if not supplier_ids:
		return {}
	rows = frappe.get_all(
		"Supplier",
		filters={"name": ["in", supplier_ids]},
		fields=["name", "supplier_name"],
	)
	return {row.name: row.supplier_name for row in rows}
